File: VodArchiver/VodArchiver/UserInfo/YoutubePlaylistUserInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;
using VodArchiver.VideoInfo;

namespace VodArchiver.UserInfo {
	public class YoutubePlaylistUserInfo : IUserInfo {
		public override ServiceVideoCategoryType Type => ServiceVideoCategoryType.YoutubePlaylist;
		public override string UserIdentifier => Playlist;
		public override bool Persistable { get => _Persistable; set => _Persistable = value; }
		public override bool AutoDownload { get => _AutoDownload; set => _AutoDownload = value; }
		public override DateTime LastRefreshedOn { get => _LastRefreshedOn; set => _LastRefreshedOn = value; }

		private bool _Persistable;
		private bool _AutoDownload;
		private DateTime _LastRefreshedOn = Util.DateTimeFromUnixTime( 0 );

		private string Playlist;

		public YoutubePlaylistUserInfo( string playlist ) {
			if ( playlist.StartsWith( "http://" ) || playlist.StartsWith( "https://" ) ) {
				Playlist = Util.GetParameterFromUri( new Uri( playlist ), "list" );
			} else {
				Playlist = playlist;
			}
			_Persistable = false;
			_AutoDownload = false;
		}

		public YoutubePlaylistUserInfo( XmlNode node ) {
			_AutoDownload = node.Attributes["autoDownload"].Value == "true";
			_LastRefreshedOn = Util.DateTimeFromUnixTime( ulong.Parse( node.Attributes["lastRefreshedOn"].Value ) );
			Playlist = node.Attributes["playlist"].Value;
			_Persistable = true;
		}

		public override XmlNode Serialize( XmlDocument document, XmlNode node ) {
			node.AppendAttribute( document, "_type", Type.ToString() );
			node.AppendAttribute( document, "autoDownload", AutoDownload ? "true" : "false" );
			node.AppendAttribute( document, "lastRefreshedOn", LastRefreshedOn.DateTimeToUnixTime().ToString() );
			node.AppendAttribute( document, "playlist", Playlist.ToString() );
			return node;
		}

		public override async Task<FetchReturnValue> Fetch( int offset, bool flat ) {
			List<IVideoInfo> videosToAdd = new List<IVideoInfo>();
			bool hasMore = true;
			long maxVideos = -1;
			int currentVideos = -1;

			List<IVideoInfo> playlistVideos = await Youtube.RetrieveVideosFromPlaylist( Playlist, flat );
			hasMore = false;
			foreach ( var v in playlistVideos ) {
				videosToAdd.Add( v );
			}
			currentVideos = playlistVideos.Count;

			if ( videosToAdd.Count <= 0 ) {
				return new FetchReturnValue { Success = true, HasMore = false, TotalVideos = maxVideos, VideoCountThisFetch = 0, Videos = videosToAdd };
			}

			return new FetchReturnValue { Success = true, HasMore = hasMore, TotalVideos = maxVideos, VideoCountThisFetch = currentVideos, Videos = videosToAdd };
		}
	}
}

File: native/vodarchiver/userinfo/youtube-playlist-user-info.cpp
#include "youtube-playlist-user-info.h"

#include <cstdint>
#include <format>
#include <memory>
#include <optional>
#include <string>
#include <utility>
#include <vector>

#include "vodarchiver/youtube_util.h"

namespace VodArchiver {
YoutubePlaylistUserInfo::YoutubePlaylistUserInfo() = default;

YoutubePlaylistUserInfo::YoutubePlaylistUserInfo(std::string playlist)
  : Playlist(std::move(playlist)) {}

YoutubePlaylistUserInfo::~YoutubePlaylistUserInfo() = default;

ServiceVideoCategoryType YoutubePlaylistUserInfo::GetType() {
    return ServiceVideoCategoryType::YoutubePlaylist;
}

std::string YoutubePlaylistUserInfo::GetUserIdentifier() {
    return Playlist;
}

FetchReturnValue YoutubePlaylistUserInfo::Fetch(JobConfig& jobConfig, size_t offset, bool flat) {
    std::vector<std::unique_ptr<IVideoInfo>> videosToAdd;
    bool hasMore = true;
    int64_t maxVideos = -1;
    int64_t currentVideos = -1;

    auto videos = Youtube::RetrieveVideosFromPlaylist(Playlist, flat, Comment);
    if (!videos.has_value()) {
        return FetchReturnValue{.Success = false, .HasMore = false};
    }
    hasMore = false;
    currentVideos = videos->size();
    for (auto& m : *videos) {
        videosToAdd.push_back(std::move(m));
    }

    if (videosToAdd.size() <= 0) {
        return FetchReturnValue{.Success = true,
                                .HasMore = false,
                                .TotalVideos = maxVideos,
                                .VideoCountThisFetch = 0,
                                .Videos = std::move(videosToAdd)};
    }

    return FetchReturnValue{.Success = true,
                            .HasMore = hasMore,
                            .TotalVideos = maxVideos,
                            .VideoCountThisFetch = currentVideos,
                            .Videos = std::move(videosToAdd)};
}

std::string YoutubePlaylistUserInfo::ToString() {
    if (!Comment.empty()) {
        return std::format(
            "{}: [{}] {}",
            ServiceVideoCategoryTypeToString(ServiceVideoCategoryType::YoutubePlaylist),
            Comment,
            Playlist);
    }
    return IUserInfo::ToString();
}

std::unique_ptr<IUserInfo> YoutubePlaylistUserInfo::Clone() const {
    auto u = std::make_unique<YoutubePlaylistUserInfo>();
    u->Persistable = this->Persistable;
    u->AutoDownload = this->AutoDownload;
    u->LastRefreshedOn = this->LastRefreshedOn;
    u->Playlist = this->Playlist;
    u->Comment = this->Comment;
    return u;
}
} // namespace VodArchiver
